File: Ex1309_MergeSortOnLinkedList/MergeSort.cpp
#include <iostream>
#include <vector>

using namespace std;

typedef int datatype;

struct node
{
	datatype data;
	struct node* next;
};
typedef struct node* PtrToNode;
typedef PtrToNode List, Position;

int IsEmpty(const List L);
int IsLast(const Position P);
List MakeList(void);
void DeleteList(List& L);
void ShowList(const List L);
void Insert(datatype x, List L, Position P);
List BuildList(const vector<datatype>& input_array);
int GetLength(const List L);

void MergeSort(List L);
void MergePass(vector<Position>& Ptrs
	, int k, int length);
void Merge(vector<Position>& Ptrs
	, int begin1, int begin2);

int main(void)
{
	vector<datatype> input = { 32,59,21,12,45,89
		,63,34,26,7,6,18,24,45 };

	List L = BuildList(input);

	cout << "Origin list:" << endl;
	ShowList(L);

	MergeSort(L);

	cout << endl << "After sorting:" << endl;
	ShowList(L);

	DeleteList(L);

	return 0;
}

int IsEmpty(const List L)
{
	return L->next == NULL;
}

int IsLast(const Position P)
{
	return P->next == NULL;
}

// Generate a new empty list
List MakeList(void)
{
	List L = new node;
	L->data = 0;
	L->next = NULL;

	return L;
}

// Delete list
void DeleteList(List& L)
{
	Position P, tmp;
	P = L;
	L = NULL;
	while (P != NULL)
	{
		tmp = P->next;
		delete P;
		P = tmp;
	}
}

void ShowList(const List L)
{
	Position P = L->next;
	while (P != NULL)
	{
		printf("%d, ", P->data);
		P = P->next;
	}
	printf("\n");
}

// Insert a node after position p
void Insert(datatype x, List L, Position P)
{
	Position TmpCell;

	TmpCell = new node;

	TmpCell->data = x;
	TmpCell->next = P->next;
	P->next = TmpCell;
}

List BuildList(const vector<datatype>& input_array)
{
	List L = MakeList();
	Position last = L;
	for (auto& i: input_array)
	{
		Insert(i, L, last);
		last = last->next;
	}
	return L;
}

int GetLength(const List L)
{
	int l = 0;
	Position P = L->next;

	while (P != NULL)
	{
		l++;
		P = P->next;
	}

	return l;
}

void MergeSort(List L)
{
	int k = 1;
	int length = GetLength(L);

	// Auxiliary array to store the address of each element
	// The index begin from 1
	vector<Position> Ptrs;
	Ptrs.push_back(NULL);

	Position P = L->next;
	L->next = NULL;
	Position P_temp = P;
	while (P != NULL)
	{
		Ptrs.push_back(P);
		P_temp = P->next;
		P->next = NULL;
		P = P_temp;
	}

	while (k < length)
	{
		MergePass(Ptrs, k, length);
		k *= 2;
	}

	L->next = Ptrs[1];
}

void MergePass(vector<Position>& Ptrs
	, int k, int length)
{
	int i = 1;
	while (i <= length - 2 * k + 1)
	{
		Merge(Ptrs, i, i + k);
		i += 2 * k;
	}
	if (i < length - k + 1)
	{
		Merge(Ptrs, i, i + k);
	}
}

void Merge(vector<Position>& Ptrs
	, int begin1, int begin2)
{
	Position former = NULL;
	Position P1, P2;
	Position P;

	P1 = Ptrs[begin1];
	P2 = Ptrs[begin2];

	if (P1->data <= P2->data)
	{
		P = former = P1;
		P1 = P1->next;
	}
	else
	{
		P = former = P2;
		P2 = P2->next;
	}
	
	while (P1 != NULL && P2 != NULL)
	{
		if (P1->data <= P2->data)
		{
			former->next = P1;
			former = P1;
			P1 = P1->next;
		}
		else
		{
			former->next = P2;
			former = P2;
			P2 = P2->next;
		}
	}

	while (P1 != NULL)
	{
		former->next = P1;
		former = P1;
		P1 = P1->next;
	}
	while (P2 != NULL)
	{
		former->next = P2;
		former = P2;
		P2 = P2->next;
	}

	// Update vector Ptrs
	int i = begin1;
	while (P != NULL)
	{
		Ptrs[i] = P;
		P = P->next;
		i++;
	}
}

File: Ex1309_MergeSortOnLinkedList/MergeSort.java
public class MergeSort {
    public static void main(String[] args) {
        MyLinkedListWithSort<Integer> list
                = new MyLinkedListWithSort<>(
                        new Integer[]{ 32,59,21,12,45,89,63
                                ,34,26,7,6,18,24,45 });

        System.out.println("Original list:");
        list.show();

        list.mergeSort();

        System.out.println("After sorting:");
        list.show();
    }
}

class MyLinkedList<E> {
    public static class Node<E> {
        public final E data;
        public Node<E> next;

        public Node(E data) {
            this(data, null);
        }

        public Node(E data, Node<E> next) {
            this.data = data;
            this.next = next;
        }

        public boolean isLast() {
            return next == null;
        }
    }

    private final Node<E> head = new Node<>(null);

    public MyLinkedList() {}

    public MyLinkedList(E[] array) {
        this();
        insertToTail(array);
    }

    public Node<E> getHead() {
        return head;
    }

    public Node<E> getFirst() {
        return getHead().next;
    }

    public boolean isEmpty() {
        return getHead().isLast();
    }

    // Insert a node after position p
    public void insert(E data, Node<E> position) {
        if (position == null) {
            throw new NullPointerException();
        }
        position.next = new Node<>(data, position.next);
    }

    public Node<E> getTail() {
        Node<E> node = getHead();
        while (!node.isLast()) {
            node = node.next;
        }
        return node;
    }

    public void insertToTail(E[] array) {
        if (array == null) {
            throw new NullPointerException();
        }

        Node<E> last = getTail();
        for (E data: array) {
            insert(data, last);
            last = last.next;
        }
    }

    public int getLength() {
        int length = 0;
        Node<E> node = getHead();
        while (!node.isLast()) {
            length++;
            node = node.next;
        }
        return length;
    }

    public void show() {
        Node<E> node = getHead();
        while (!node.isLast()) {
            node = node.next;
            System.out.print(node.data + ", ");
        }
        System.out.println();
    }
}

class MyLinkedListWithSort<E extends Comparable<E>>
        extends MyLinkedList<E> {
    public MyLinkedListWithSort() {
        super();
    }

    public MyLinkedListWithSort(E[] array) {
        super(array);
    }

    public void mergeSort() {
        int k = 1;
        int length = getLength();

        // Auxiliary array to store the address of each element
        // The index begin from 1
        @SuppressWarnings("unchecked")
        Node<E>[] nodes = (Node<E>[])new Node[length + 1];

        Node<E> node = getFirst();
        getHead().next = null;
        Node<E> temp = node;
        while (node != null) {
            nodes[k] = node;
            k++;

            temp = node.next;
            node.next = null;
            node = temp;
        }

        k = 1;
        while (k < length) {
            mergePass(nodes, k, length);
            k *= 2;
        }

        getHead().next = nodes[1];
    }

    private void mergePass(Node<E>[] nodes, int k, int length) {
        int i;
        for (i = 1; i <= length - 2 * k + 1; i += 2 * k) {
            merge(nodes, i, i + k);
        }
        if (i < length - k + 1) {
            merge(nodes, i, i + k);
        }
    }

    private void merge(Node<E>[] nodes, int begin1, int begin2) {
        Node<E> former = null;
        Node<E> p1, p2;
        Node<E> p;

        p1 = nodes[begin1];
        p2 = nodes[begin2];

        if (p1.data.compareTo(p2.data) <= 0) {
            p = former = p1;
            p1 = p1.next;
        }
        else {
            p = former = p2;
            p2 = p2.next;
        }

        while (p1 != null && p2 != null) {
            if (p1.data.compareTo(p2.data) <= 0) {
                former.next = p1;
                former = p1;
                p1 = p1.next;
            }
            else {
                former.next = p2;
                former = p2;
                p2 = p2.next;
            }
        }

        while (p1 != null) {
            former.next = p1;
            former = p1;
            p1 = p1.next;
        }
        while (p2 != null) {
            former.next = p2;
            former = p2;
            p2 = p2.next;
        }

        // Update nodes
        int i = begin1;
        while (p != null) {
            nodes[i] = p;
            p = p.next;
            i++;
        }
    }
}
